### PR TITLE
Motion constraint

### PR DESCRIPTION
**Pull request by jafu888**

### CODE REVIEW

**Comment by jafu888** [comment]
This support a motion subsection of ConstraintSets 

    csetName: {
       idToConstrain : {
           motion: {
              pathArc : 'startVertical' // move this one view in an arc
              relativeTo: 'id'           // move relative TBD
              easing: 'curve'         //  move individual with an easing curve 
              stagger: 2               // set the stagger value  (and matching stagger attribute in Transition
              quantize: steps or [steps, 'interpolator' phase ] // move in discrete steps 
           }
       }
     }

https://user-images.githubusercontent.com/15019413/168735172-849b0302-8a82-4f62-95c0-f6e81c1c0ee4.mp4


     Quick demo of the capabilities.

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/motion/Motion.java [hunk 1]
```diff
@@ -1727,15 +1735,20 @@ public boolean setValue(int id, int value) {
 
     @Override
     public boolean setValue(int id, float value) {
+        if (MotionType.TYPE_QUANTIZE_MOTION_PHASE == id) {
+            mQuantizeMotionPhase =  value;
+        } else if (MotionType.TYPE_STAGGER == id) {
+            mMotionStagger = value;
+        }
```
Comment (oscar-ad) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/motion/Motion.java:1742]: shouldn't we return true in these cases?

Comment (oscar-ad) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/motion/Motion.java:1742]: shouldn't we return true in these cases?

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/motion/Motion.java [hunk 2]
```diff
@@ -1727,15 +1735,20 @@ public boolean setValue(int id, int value) {
...
         return false;
     }
 
     @Override
     public boolean setValue(int id, String value) {
-        if (TransitionType.TYPE_INTERPOLATOR == id) {
-            System.out.println("TYPE_INTERPOLATOR  " + value);
+        if (TransitionType.TYPE_INTERPOLATOR == id || MotionType.TYPE_QUANTIZE_INTERPOLATOR_TYPE == id) {
```
Comment (oscar-ad) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/motion/Motion.java:1748]: Same as above, should we return true?

Comment (oscar-ad) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/motion/Motion.java:1748]: Same as above, should we return true?

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/motion/Motion.java [hunk 3]
```diff
@@ -1748,4 +1761,27 @@ public boolean setValue(int id, boolean value) {
...
+     */
+    public void setStaggerScale(float staggerScale) {
+        mStaggerScale = staggerScale;
+    }
+
+    /**
+     * set the offset used in calculating stagger launches
+     * @param staggerOffset
+     */
+    public void setStaggerOffset(float staggerOffset) {
```
Comment (oscar-ad) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/motion/Motion.java:1776]: What is this offset? What unit is this supposed to represent?

Comment (oscar-ad) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/motion/Motion.java:1776]: What is this offset? What unit is this supposed to represent?

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/motion/Motion.java [hunk 4]
```diff
@@ -1748,4 +1761,27 @@ public boolean setValue(int id, boolean value) {
...
+        mStaggerOffset = staggerOffset;
+    }
+
+    /**
+     * Get the motionStagger valule if they are not using positions
+     * @return
+     */
+    public float getMotionStagger() {
+        return mMotionStagger;
```
Comment (oscar-ad) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/motion/Motion.java:1785]: It seems that we return the stagger value regardless of positions, is the javadoc wrong or the implementation?

Comment (oscar-ad) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/motion/Motion.java:1785]: It seems that we return the stagger value regardless of positions, is the javadoc wrong or the implementation?

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/motion/MotionWidget.java [hunk 5]
```diff
@@ -148,10 +148,21 @@ public MotionWidget(WidgetFrame f) {
...
+    }
     @Override
     public boolean setValue(int id, int value) {
-        return setValueAttributes(id, value);
-    }
+        boolean set = setValueAttributes(id, value);
+        if (set) {
+            return true;
+        }
+        return setValueMotion(id, value);    }
```
Comment (oscar-ad) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/motion/MotionWidget.java:165]: Please have the closing bracket `}` on a new line.

Comment (oscar-ad) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/motion/MotionWidget.java:165]: Please have the closing bracket `}` on a new line.

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/ConstraintSetParser.java [hunk 6]
```diff
@@ -1012,6 +1022,93 @@ static void parseCustomProperties(
 
     }
 
+    private static int map(String val, String... types) {
```
Comment (oscar-ad) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/ConstraintSetParser.java:1025]: `indexOf` ? Imo map implies we are mapping the value to something.

Comment (oscar-ad) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/ConstraintSetParser.java:1025]: `indexOf` ? Imo map implies we are mapping the value to something.

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/Transition.java [hunk 7]
```diff
@@ -984,4 +987,85 @@ static class KeyPosition {
...
+                WidgetState widgetState = mState.get(widgetId);
+                Motion f = widgetState.mMotionControl;
+                float x = f.getFinalX();
+                float y = f.getFinalY();
+                float widgetStagger = x + y;
+                min = Math.min(min, widgetStagger);
+                max = Math.max(max, widgetStagger);
+            }
+
+            Utils.log("min " + min + " max " + max);
```
Comment (oscar-ad) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/Transition.java:1050]: If (DEBUG) { ... }

Comment (oscar-ad) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/Transition.java:1050]: If (DEBUG) { ... }

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/Transition.java [hunk 8]
```diff
@@ -984,4 +987,85 @@ static class KeyPosition {
...
+                float y = f.getFinalY();
+                float widgetStagger = x + y;
+                float offset = stagger - stagger * (widgetStagger - (min)) / (max - (min));
+                if (flip) {
+                    offset = stagger - stagger
+                            * ((max - widgetStagger) / (max - min));
+                }
+
+                float scale = 1 / (1 - stagger);
+                Utils.log("1 / (1 - mStagger) =  " + scale + " offset= " + offset);
```
Comment (oscar-ad) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/Transition.java:1065]: if (DEBUG) { ... }

Comment (oscar-ad) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/Transition.java:1065]: if (DEBUG) { ... }

**Review by oscar-ad — CHANGES_REQUESTED** [review]

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/motion/Motion.java (hunk 1 above)
Comment (jafu888) [reply constraintlayout/core/src/main/java/androidx/constraintlayout/core/motion/Motion.java:1742]: It does not matter but to be consistent change

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/motion/Motion.java (hunk 2 above)
Comment (jafu888) [reply constraintlayout/core/src/main/java/androidx/constraintlayout/core/motion/Motion.java:1748]: same

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/Transition.java (hunk 7 above)
Comment (jafu888) [reply constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/Transition.java:1050]: deleted

Comment (jafu888) [reply constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/Transition.java:1050]: deleted

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/Transition.java (hunk 8 above)
Comment (jafu888) [reply constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/Transition.java:1065]: deleted

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/motion/Motion.java (hunk 1 above)
Comment (jafu888) [reply constraintlayout/core/src/main/java/androidx/constraintlayout/core/motion/Motion.java]: done

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/Transition.java [hunk 9]
```diff
@@ -833,7 +834,8 @@ private WidgetState getWidgetState(String widgetId,
         if (widgetState == null) {
             widgetState = new WidgetState();
             mBundle.applyDelta(widgetState.mMotionControl);
-
+            widgetState.mMotionWidgetStart.updateMotion( widgetState.mMotionWidgetStart);
```
Comment (oscar-ad) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/Transition.java:837]: Can you double check this? Seems unnecessary

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/WidgetFrame.java [hunk 10]
```diff
@@ -18,7 +18,10 @@
 
 import androidx.constraintlayout.core.motion.CustomAttribute;
 import androidx.constraintlayout.core.motion.CustomVariable;
+import androidx.constraintlayout.core.motion.Motion;
+import androidx.constraintlayout.core.motion.utils.TypedBundle;
 import androidx.constraintlayout.core.motion.utils.TypedValues;
+import androidx.constraintlayout.core.motion.utils.Utils;
```
Comment (oscar-ad) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/WidgetFrame.java:24]: Double check with optimize imports, seems unused now.

Comment (jafu888) [reply constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/WidgetFrame.java:24]: removed

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/Transition.java (hunk 9 above)
Comment (jafu888) [reply constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/Transition.java:837]: removed

**Review by oscar-ad — APPROVED** [review]

**Pull request merged**